packages/react-devtools-extensions/src/injectGlobalHook.js

```diff
@@ -89,13 +73,10 @@ if (sessionStorageGetItem(SESSION_STORAGE_RELOAD_AND_PROFILE_KEY) === 'true') {
 // Inject a __REACT_DEVTOOLS_GLOBAL_HOOK__ global for React to interact with.
 // Only do this for HTML documents though, to avoid e.g. breaking syntax highlighting for XML docs.
 if ('text/html' === document.contentType) {
-  injectCode(
-    ';(' +
-      installHook.toString() +
-      '(window))' +
-      saveNativeValues +
-      detectReact,
-  );
+  const script = document.createElement('script');
+  script.src = chrome.runtime.getURL('build/extensionHook.js');
+  document.documentElement.appendChild(script);
+  script.parentNode.removeChild(script);
```

Comment:
This seems like it would change the _timing_ of the hook being injected into the page?

The reason we were using `toString` before was b'c it let us _synchronously_ inject the script. (The hook needs to be there before react-dom or any other react renderer runs, or the renderer won't register itself with DevTools).

Reply:
If we are inserting before the DOM is ready I think this is still a sync load. I can test this out though to ensure that this loads before any other code that we expect. Which file loads react-dom ?

Reply:
> If we are inserting before the DOM is ready I think this is still a sync load since we `"run_at": "document_start"` this gets injected before any other code still. I can test this out though to ensure that this loads before any other code that we expect.

Yes, please! That would be great to know!

>  Which file loads react-dom ? or are you talking about the use of the extension loading react-dom?

Depends on the website :)

Reply:
I think you can test it more easily than that by just adding an inline script tag?

```js
<head>
<script type= text/javascript">
  console.log("__REACT_DEVTOOLS_GLOBAL_HOOK__:", typeof window.__REACT_DEVTOOLS_GLOBAL_HOOK__);
</script>
</head>
```

Reply:
> The hook needs to be there before react-dom or any other react renderer runs

earlier you said this needs to be the case but ive found even when its not the dev tools work as expected

Reply:
```html
<html>
  <body>
    <div id="root"></div>
    <script crossorigin src="https://unpkg.com/react@17/umd/react.production.min.js"></script>
    <script crossorigin src="https://unpkg.com/react-dom@17/umd/react-dom.production.min.js"></script>
    <script crossorigin src="https://unpkg.com/@babel/standalone/babel.min.js"></script>

    <script type="text/babel">
        console.log("__REACT_DEVTOOLS_GLOBAL_HOOK__:", typeof window.__REACT_DEVTOOLS_GLOBAL_HOOK__);
        class Hello extends React.Component{
          render() {
            return <p>Hello {this.props.name}</p>;
          };
        }

        ReactDOM.render(
          <Hello name='World'/>,
          document.getElementById('root'),
        );
    </script>
  </body>
</html>
```
here is a pretty basic example that logs `typeof window.__REACT_DEVTOOLS_GLOBAL_HOOK__` as  `object`

Reply:
![image](https://user-images.githubusercontent.com/745064/129160157-b2c47147-044a-4001-b679-58d2dea6a5e1.png)

Reply:
I also created a [simple test server](https://github.com/colegillespie/react-extension-test) with some test code that compiles react into a single shipable file to see if that made a different vs loading from cdn. both work as expected

Reply:
> > The hook needs to be there before react-dom or any other react renderer runs
>
> earlier you said this needs to be the case but ive found even when its not the case dev tools still work as expected

Can you elaborate on this? This makes me concerned that the way you're testing is giving a false positive.

React DOM (and other renderers) only attach themselves to the DevTools hook if it's present when the module is loaded.

This value is initialized during module initialization:
https://github.com/facebook/react/blob/07e58658c9bb182eab436c5fe9e1e66078ea0b19/packages/react-reconciler/src/ReactFiberDevToolsHook.new.js#L36-L37

Based on that value, React internals will switch on or off important functionality that DevTools depends on.

Similarly, React DOM injects itself during initialization:
https://github.com/facebook/react/blob/07e58658c9bb182eab436c5fe9e1e66078ea0b19/packages/react-dom/src/client/ReactDOM.js#L191-L196

So if the DevTools hook isn't around, the connection will be missed until the page is reloaded (and the module is re-evaluated)

Reply:
this is great information, let me set some breakpoints and test locally and then will elaborate more on the findings. Thanks for working with me on this.